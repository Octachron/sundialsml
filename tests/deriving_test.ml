open Pprint
open Camlp4.PreCast
open Expr_of

type 'a poly1 = 'a
and  ('a,'b) poly2 = 'a * 'b
and  ('a,'b,'c) poly3 = 'a
and  ('a,'b,'c,'d) poly4 = 'a * int * 'c * unit
and  record1 = { mutable mfield11 : unit;
                 field12: unit; }
and 'a poly1_record2 = { opt_field21 : 'a;
                         mutable opt_mfield22 : 'a array;
                         mutable opt_mfield23 : 'a Lazy.t;
                       }
and ('a,'b) poly2_record3 =
  { polyrec_field31 : ('a,int) poly2_record3 list lazy_t;
    backref_field32 : ('a,'b) poly2;
    fwdref_field33 : var1_0
  }
and var1_0 = OptVar11
and var2_00 = Var21 | Var22
and var3_012 = Var31 | OptVar32 of unit | Var33 of unit * unit
and var4_pair = Var41 of (unit * unit)
and 'a poly1_var5_12 = Var51 of 'a | Var52 of 'a * 'a
and ('a,'b) poly2_var6_123 = Var61 of 'a | Var62 of 'a * 'b
                           | OptVar63 of 'a * 'b * 'b * 'b
and ('a,'b,'c) poly3_var7_2 = Var71 of ('a * 'b) * 'c
deriving
  (expr_of ~prefix:Deriving_test
           ~alias:(Lazy.t = lazy_t)
  ,pretty ~prefix:Deriving_test
          ~alias:(Lazy.t = lazy_t)
          ~optional:(Deriving_test,
                     opt_field21,
                     opt_mfield22,
                     opt_mfield23,
                     OptVar11,     (* <- this should have no effect *)
                     OptVar32,
                     OptVar63)
)

module Camlp4aux = Camlp4aux.Make (Syntax)
let test title expr_of pp x =
  Format.printf "<%s>@\nRWI:    " title;
  with_read_write_invariance (fun () -> pp Format.std_formatter x);
  Format.printf "@\nNo RWI: ";
  pp Format.std_formatter x;
  Format.printf "@\nexpr:   ";
  Camlp4aux.ocaml_printer#expr Format.std_formatter (expr_of x);
  Format.printf "@\n@\n"
in
test "poly1" (expr_of_poly1 expr_of_unit) (pp_poly1 pp_unit) ();
test "poly1" (expr_of_poly1 expr_of_int) (pp_poly1 pp_int) (-1);
test "poly2" (expr_of_poly2 expr_of_int expr_of_int)
  (pp_poly2 pp_int pp_int) (-1, -1);
test "poly2" (expr_of_poly2 expr_of_int expr_of_int)
  (pp_poly2 pp_int pp_int) (1, 1);
test "poly3" (expr_of_poly3 expr_of_unit expr_of_int expr_of_int)
  (pp_poly3 pp_unit pp_int pp_int) ();
test "poly4"
  (expr_of_poly4 expr_of_unit expr_of_unit expr_of_unit expr_of_unit)
  (pp_poly4 pp_unit pp_unit pp_unit pp_unit)
  ((),-1,(),());
test "poly4"
  (expr_of_poly4 expr_of_int expr_of_int expr_of_int expr_of_int)
  (pp_poly4 pp_int pp_int pp_int pp_int)
  (-1,-1,-1,());
test "record1" expr_of_record1 pp_record1 { mfield11 = (); field12 = () };
test "poly1_record2"
  (expr_of_poly1_record2 expr_of_unit)
  (pp_poly1_record2 pp_unit)
  { opt_field21 = (); opt_mfield22 = [| (); () |];
    opt_mfield23 = lazy () };
test "poly1_record2"
  (expr_of_poly1_record2 expr_of_int)
  (pp_poly1_record2 pp_int)
  { opt_field21 = -1; opt_mfield22 = [| -1; 0 |];
    opt_mfield23 = lazy (5+3) };
test "poly2_record3"
  (expr_of_poly2_record3 expr_of_unit expr_of_unit)
  (pp_poly2_record3 pp_unit pp_unit)
  { polyrec_field31 = lazy [];
    backref_field32 = (), ();
    fwdref_field33 = OptVar11 };
test "poly2_record3"
  (expr_of_poly2_record3 expr_of_int expr_of_bool)
  (pp_poly2_record3 pp_int pp_bool)
  { polyrec_field31 = Lazy.lazy_from_val
        [{ polyrec_field31 = lazy [];
           backref_field32 = (1, 2);
           fwdref_field33 = OptVar11 }];
    backref_field32 = -1, false;
    fwdref_field33 = OptVar11 };
test "var1_0" expr_of_var1_0 pp_var1_0 OptVar11;
test "var2_00"
  (expr_of_list expr_of_var2_00)
  (pp_list pp_var2_00)
  [Var21; Var22];
test "var3_012"
  (expr_of_list expr_of_var3_012)
  (pp_list pp_var3_012)
  [Var31; OptVar32 (); Var33 ((), ())];
test "var4_pair" expr_of_var4_pair pp_var4_pair (Var41 ((), ()));
test "poly1_var5_12"
  (expr_of_list (expr_of_poly1_var5_12 expr_of_unit))
  (pp_list (pp_poly1_var5_12 pp_unit))
  [Var51 (); Var52 ((), ())];
test "poly1_var5_12"
  (expr_of_list (expr_of_poly1_var5_12 expr_of_int))
  (pp_list (pp_poly1_var5_12 pp_int))
  [Var51 0; Var51 (-1); Var52 (0, -1)];
test "poly2_var6_123"
  (expr_of_list (expr_of_poly2_var6_123 expr_of_unit expr_of_unit))
  (pp_list (pp_poly2_var6_123 pp_unit pp_unit))
  [Var61 (); Var62 ((), ()); OptVar63 ((), (), (), ())];
test "poly2_var6_123"
  (expr_of_list (expr_of_poly2_var6_123 expr_of_int expr_of_bool))
  (pp_list (pp_poly2_var6_123 pp_int pp_bool))
  [Var61 1; Var61 (-1); Var62 (0, true); OptVar63 (0, true, false, true)];
test "poly3_var7_2"
  (expr_of_list (expr_of_poly3_var7_2 expr_of_unit expr_of_unit expr_of_unit))
  (pp_list (pp_poly3_var7_2 pp_unit pp_unit pp_unit))
  [Var71 (((), ()), ())];
test "poly3_var7_2"
  (expr_of_list (expr_of_poly3_var7_2 expr_of_int expr_of_bool expr_of_char))
  (pp_list (pp_poly3_var7_2 pp_int pp_bool pp_char))
  [Var71 ((-1, true), 'a')]
