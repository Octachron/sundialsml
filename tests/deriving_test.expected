<poly1>
RWI:    ()
No RWI: ()
expr:   ()

<poly1>
RWI:    -1
No RWI: -1
expr:   (-1)

<poly2>
RWI:    (-1, -1)
No RWI: (-1, -1)
expr:   ((-1), (-1))

<poly2>
RWI:    (1, 1)
No RWI: (1, 1)
expr:   (1, 1)

<poly3>
RWI:    ()
No RWI: ()
expr:   ()

<poly4>
RWI:    ((), -1, (), ())
No RWI: ((), -1, (), ())
expr:   ((), (-1), (), ())

<poly4>
RWI:    (-1, -1, -1, ())
No RWI: (-1, -1, -1, ())
expr:   ((-1), (-1), (-1), ())

<record1>
RWI:    {Deriving_test.mfield11 = (); Deriving_test.field12 = ()}
No RWI: {mfield11 = (); field12 = ()}
expr:   { Deriving_test.mfield11 = (); Deriving_test.field12 = (); }

<poly1_record2>
RWI:    {Deriving_test.opt_field21 = ();
         Deriving_test.opt_mfield22 = [|(); ()|];
         Deriving_test.opt_mfield23 = lazy ()}
No RWI: {}
expr:   {
          Deriving_test.opt_field21 = ();
          Deriving_test.opt_mfield22 = [| (); () |];
          Deriving_test.opt_mfield23 = lazy ();
        }

<poly1_record2>
RWI:    {Deriving_test.opt_field21 = -1;
         Deriving_test.opt_mfield22 = [|-1; 0|];
         Deriving_test.opt_mfield23 = <lazy>}
No RWI: {}
expr:   {
          Deriving_test.opt_field21 = (-1);
          Deriving_test.opt_mfield22 = [| (-1); 0 |];
          Deriving_test.opt_mfield23 = lazy 8;
        }

<poly2_record3>
RWI:    {Deriving_test.polyrec_field31 = lazy [];
         Deriving_test.backref_field32 = ((), ());
         Deriving_test.fwdref_field33 = OptVar11}
No RWI: {polyrec_field31 = lazy [];
         backref_field32 = ((), ());
         fwdref_field33 = OptVar11}
expr:   {
          Deriving_test.polyrec_field31 = lazy [];
          Deriving_test.backref_field32 = ((), ());
          Deriving_test.fwdref_field33 = Deriving_test.OptVar11;
        }

<poly2_record3>
RWI:    {Deriving_test.polyrec_field31 = lazy [{Deriving_test.polyrec_field31 =
                                                 lazy [];
                                                Deriving_test.backref_field32 =
                                                 (1, 2);
                                                Deriving_test.fwdref_field33 =
                                                 OptVar11}];
         Deriving_test.backref_field32 = (-1, false);
         Deriving_test.fwdref_field33 = OptVar11}
No RWI: {polyrec_field31 = lazy [{polyrec_field31 = lazy [];
                                  backref_field32 = (1, 2);
                                  fwdref_field33 = OptVar11}];
         backref_field32 = (-1, false);
         fwdref_field33 = OptVar11}
expr:   {
          Deriving_test.polyrec_field31 =
            lazy
              [ {
                  Deriving_test.polyrec_field31 = lazy [];
                  Deriving_test.backref_field32 = (1, 2);
                  Deriving_test.fwdref_field33 = Deriving_test.OptVar11;
                } ];
          Deriving_test.backref_field32 = ((-1), false);
          Deriving_test.fwdref_field33 = Deriving_test.OptVar11;
        }

<var1_0>
RWI:    OptVar11
No RWI: OptVar11
expr:   Deriving_test.
OptVar11

<var2_00>
RWI:    [Var21; Var22]
No RWI: [Var21; Var22]
expr:   [ Deriving_test.Var21; Deriving_test.Var22 ]

<var3_012>
RWI:    [Var31; Deriving_test.OptVar32 (); Deriving_test.Var33 ((), ())]
No RWI: [Var31; (); Var33 ((), ())]
expr:   [ Deriving_test.Var31; Deriving_test.OptVar32 ();
          Deriving_test.Var33 ((), ()) ]

<var4_pair>
RWI:    Deriving_test.Var41 ((), ())
No RWI: Var41 ((), ())
expr:   Deriving_test.Var41 (((), ()))

<poly1_var5_12>
RWI:    [Deriving_test.Var51 (); Deriving_test.Var52 ((), ())]
No RWI: [Var51 (); Var52 ((), ())]
expr:   [ Deriving_test.Var51 (); Deriving_test.Var52 ((), ()) ]

<poly1_var5_12>
RWI:    [Deriving_test.Var51 0; Deriving_test.Var51 (-1);
         Deriving_test.Var52 (0, -1)]
No RWI: [Var51 0; Var51 (-1); Var52 (0, -1)]
expr:   [ Deriving_test.Var51 0; Deriving_test.Var51 (-1);
          Deriving_test.Var52 (0, (-1)) ]

<poly2_var6_123>
RWI:    [Deriving_test.Var61 (); Deriving_test.Var62 ((), ());
         Deriving_test.OptVar63 ((), (), (), ())]
No RWI: [Var61 (); Var62 ((), ()); ((), (), (), ())]
expr:   [ Deriving_test.Var61 (); Deriving_test.Var62 ((), ());
          Deriving_test.OptVar63 ((), (), (), ()) ]

<poly2_var6_123>
RWI:    [Deriving_test.Var61 1; Deriving_test.Var61 (-1);
         Deriving_test.Var62 (0, true);
         Deriving_test.OptVar63 (0, true, false, true)]
No RWI: [Var61 1; Var61 (-1); Var62 (0, true); (0, true, false, true)]
expr:   [ Deriving_test.Var61 1; Deriving_test.Var61 (-1);
          Deriving_test.Var62 (0, true);
          Deriving_test.OptVar63 (0, true, false, true) ]

<poly3_var7_2>
RWI:    [Deriving_test.Var71 (((), ()), ())]
No RWI: [Var71 (((), ()), ())]
expr:   [ Deriving_test.Var71 (((), ()), ()) ]

<poly3_var7_2>
RWI:    [Deriving_test.Var71 ((-1, true), 'a')]
No RWI: [Var71 ((-1, true), 'a')]
expr:   [ Deriving_test.Var71 (((-1), true), 'a') ]

